### PR TITLE
Listen for route and location updates and assign to view model

### PR DESCRIPTION
**Pull request by LukasPaczos**
Fixes an issue where both route and location observers wouldn't be subscribed, which in turn resulted in `NavigationMapboxMap` not displaying up-to-date routes and positon. Also, unregisters the offroute listener when the navigation is stopped.

### CODE REVIEW

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java [hunk 1]
```diff
@@ -438,7 +443,14 @@ public void fasterRouteFound(DirectionsRoute directionsRoute) {
     }
   };*/
 
-  private Router.Callback routeEngineCallback = new NavigationViewRouteEngineListener(this);
```
Comment (JunDai) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java]: by removing this, should we also delete the class `NavigationViewRouteEngineListener`?

Comment (LukasPaczos) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java]: good catch! done

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java [hunk 2]
```diff
@@ -438,7 +443,14 @@ public void fasterRouteFound(DirectionsRoute directionsRoute) {
...
+  private RoutesObserver routesObserver = new RoutesObserver() {
+    @Override
+    public void onRoutesChanged(@NotNull List<? extends DirectionsRoute> routes) {
+      if (routes.size() > 0) {
+        route.setValue(routes.get(0));
+      }
+    }
+  };
```
Comment (abhishek1508) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java]: @LukasPaczos I am still trying to understand why we need the `RoutesObserver`

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java [hunk 3]
```diff
@@ -26,6 +26,7 @@
 import com.mapbox.navigation.base.typedef.TimeFormatType;
 import com.mapbox.navigation.core.MapboxDistanceFormatter;
 import com.mapbox.navigation.core.MapboxNavigation;
+import com.mapbox.navigation.core.directions.session.RoutesObserver;
```
Comment (abhishek1508) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java]: All changes in this PR except for `RoutesObserver` is already addressed in this [PR](https://github.com/mapbox/mapbox-navigation-android/pull/2527)

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java [hunk 4]
```diff
@@ -450,8 +462,7 @@ private void startNavigation(DirectionsRoute route) {
     }
   }
 
-  private void updateReplayEngine(DirectionsRoute route) {
-    locationEngineConductor.updateSimulatedRoute(route);
+  private void updateReplayEngine(DirectionsRoute route) { locationEngineConductor.updateSimulatedRoute(route);
```
Comment (Guardiola31337) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java]: NIT unnecessary formatting change?

**Comment by Guardiola31337** [comment]
Also CI is ❌ 

```
> Task :libnavigation-ui:checkstyle FAILED
[ant:checkstyle] [ERROR] /root/code/libnavigation-ui/src/main/java/com/******/navigation/ui/NavigationViewModel.java:24:8: Unused import - com.******.navigation.base.route.Router. [UnusedImports]
[ant:checkstyle] [ERROR] /root/code/libnavigation-ui/src/main/java/com/******/navigation/ui/NavigationViewModel.java:465: 'method def' child has incorrect indentation level 2, expected level should be 4. [Indentation]
[ant:checkstyle] [ERROR] /root/code/libnavigation-ui/src/main/java/com/******/navigation/ui/NavigationViewModel.java:465:58: '{' at column 58 should have line break after. [LeftCurly]

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':libnavigation-ui:checkstyle'.
> Checkstyle rule violations were found. See the report at: file:///root/code/libnavigation-ui/build/reports/checkstyle/checkstyle.html
  Checkstyle files with violations: 1
  Checkstyle violations by severity: [error:3]
```

Instructions on how to install Java Code style can be found here https://github.com/mapbox/mapbox-navigation-android/blob/master/CONTRIBUTING.md https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/codestyle-setup.md

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java (hunk 2 above)
Comment (LukasPaczos) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java]: @abhishek1508 if you step through the code, you'll notice that `NavigationMapboxMap#drawRoute` is not called from anywhere. This means that the route will not ultimately be displayed.

That's because the `route` live data object that's stored in the view model has only been updated by now-removed and previously unused `NavigationViewRouteEngineListener`.

Comment (LukasPaczos) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java]: Using the observer additionally guarantees us that the drawn route will always be the one that was passed to the Nav Native for active guidance, without having to perform any additional testing.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java [hunk 5]
```diff
@@ -246,6 +247,8 @@ void stopNavigation() {
     navigation.unregisterRouteProgressObserver(navigationViewVm);
     navigation.unregisterLocationObserver(navigationViewVm);
     navigation.unregisterVoiceInstructionsObserver(voiceInstructionsObserver);
+    navigation.unregisterOffRouteObserver(offRouteListener);
```
Comment (abhishek1508) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java]: Can we remove this line? We have it in other PR. This will cause merge conflicts.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java [hunk 6]
```diff
@@ -401,9 +404,10 @@ private void initializeNavigation(Context context, NavigationOptions options, Lo
 
   private void addNavigationListeners() {
     navigation.registerRouteProgressObserver(navigationViewVm);
-    navigation.unregisterLocationObserver(navigationViewVm);
+    navigation.registerLocationObserver(navigationViewVm);
```
Comment (abhishek1508) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java]: Can we remove this line? We have it in other PR. This will cause merge conflicts.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java (hunk 2 above)
Comment (abhishek1508) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java]: 👍

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java (hunk 5 above)
Comment (LukasPaczos) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java]: Could we merge here to unlock other work? This should be a simple fix (or maybe git will even figure it out by itself).

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java (hunk 6 above)
Comment (LukasPaczos) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java]: Same as above.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java (hunk 5 above)
Comment (JunDai) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/NavigationViewModel.java]: given it's a simple and needed fix. we might want to merge it asap. I would assume the other PR is the one to create the ExampleActivity and that PR will need more time for reviewing before merge.
